mods/beds/api.lua

```diff
@@ -44,23 +44,42 @@ function beds.register_bed(name, def)
 			fixed = def.selectionbox,
 		},
 
-		after_place_node = function(pos, placer, itemstack)
-			local n = minetest.get_node_or_nil(pos)
-			if not n or not n.param2 then
-				minetest.remove_node(pos)
-				return true
+		on_place = function(itemstack, placer, pointed_thing)
+			local under = pointed_thing.under
+			local pos
+			if minetest.registered_items[minetest.get_node(under).name].buildable_to then
+				pos = under
+			else
+				pos = pointed_thing.above
 			end
-			local dir = minetest.facedir_to_dir(n.param2)
-			local p = vector.add(pos, dir)
-			local n2 = minetest.get_node_or_nil(p)
-			local def = n2 and minetest.registered_items[n2.name]
-			if not def or not def.buildable_to then
-				minetest.remove_node(pos)
-				return true
+
+			if minetest.is_protected(pos, placer:get_player_name()) and
+					not minetest.check_player_privs(placer, "protection_bypass") then
+				minetest.record_protection_violation(pos, placer:get_player_name())
+				return itemstack
+			end
+
+			local dir = minetest.dir_to_facedir(placer:get_look_dir())
+			local botpos = vector.add(pos, minetest.facedir_to_dir(dir))
+
+			if minetest.is_protected(botpos, placer:get_player_name()) and
+					not minetest.check_player_privs(placer, "protection_bypass") then
+				minetest.record_protection_violation(botpos, placer:get_player_name())
+				return itemstack
 			end
-			minetest.set_node(p, {name = n.name:gsub("%_bottom", "_top"), param2 = n.param2})
-			return false
-		end,	
+
+			if not minetest.registered_nodes[minetest.get_node(botpos).name].buildable_to then
+				return itemstack
+			end
+
+			minetest.set_node(pos, {name = name .. "_bottom", param2 = dir})
+			minetest.set_node(botpos, {name = name .. "_top", param2 = dir})
```

Comment:
This will have the unloaded block bug (where placing something next to an unloaded block causes a node to be forcibly replaced in the block as soon as it's loaded).
You have to check if the node you're trying to place in is `"ignore"` and abort if so.
If possible you should avoid `on_place` and use `after_place_node` instead.  It's easy to miss something and have a bug like this.